File: jmix-templates/content/project/flowui-single-module-application/src/main/java/${project_rootPath}/view/user/UserDetailView.java
package ${project_rootPackage}.view.user;

import ${project_rootPackage}.entity.User;
import ${project_rootPackage}.view.main.MainView;
import com.vaadin.flow.component.combobox.ComboBox;
import com.vaadin.flow.component.textfield.PasswordField;
import com.vaadin.flow.router.Route;
import com.vaadin.flow.router.RoutePrefix;
import io.jmix.core.EntityStates;
import io.jmix.flowui.Notifications;
import io.jmix.flowui.component.textfield.TypedTextField;
import io.jmix.flowui.view.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;

import java.util.List;
import java.util.Objects;
import java.util.TimeZone;

@Route(value = ":id", layout = MainView.class)
@RoutePrefix(UserListView.ROUTE)
@UiController("User.detail")
@UiDescriptor("user-detail-view.xml")
@EditedEntityContainer("userDc")
public class UserDetailView extends StandardDetailView<User> {

    @ComponentId
    private TypedTextField<String> usernameField;
    @ComponentId
    private PasswordField passwordField;
    @ComponentId
    private PasswordField confirmPasswordField;
    @ComponentId
    private ComboBox<String> timeZoneField;

    @Autowired
    private EntityStates entityStates;
    @Autowired
    private MessageBundle messageBundle;
    @Autowired
    private Notifications notifications;
    @Autowired
    private PasswordEncoder passwordEncoder;

    @Subscribe
    public void onInit(InitEvent event) {
        timeZoneField.setItems(List.of(TimeZone.getAvailableIDs()));
    }

    @Subscribe
    public void onInitEntity(InitEntityEvent<User> event) {
        usernameField.setReadOnly(false);
        passwordField.setVisible(true);
        confirmPasswordField.setVisible(true);
    }

    @Subscribe
    public void onAfterShow(AfterShowEvent event) {
        if (entityStates.isNew(getEditedEntity())) {
            usernameField.focus();
        }
    }

    @Subscribe
    protected void onBeforeCommit(BeforeCommitChangesEvent event) {
        if (entityStates.isNew(getEditedEntity())) {
            if (!Objects.equals(passwordField.getValue(), confirmPasswordField.getValue())) {
                notifications.create(messageBundle.getMessage("passwordsDoNotMatch"))
                        .withType(Notifications.Type.WARNING)
                        .show();
                event.preventCommit();
            }
            getEditedEntity().setPassword(passwordEncoder.encode(passwordField.getValue()));
        }
    }
}

File: jmix-templates/content/project/application-kotlin/src/main/kotlin/${project_rootPath}/view/user/UserDetailView.kt
package ${project_rootPackage}.view.user

import ${project_rootPackage}.entity.User
import ${project_rootPackage}.view.main.MainView
import com.vaadin.flow.component.combobox.ComboBox
import com.vaadin.flow.component.textfield.PasswordField
import com.vaadin.flow.router.Route
import io.jmix.core.EntityStates
import io.jmix.flowui.component.textfield.TypedTextField
import io.jmix.flowui.view.*
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.crypto.password.PasswordEncoder

import java.util.*

@Route(value = "users/:id", layout = MainView::class)
@ViewController("${normalizedPrefix_underscore}User.detail")
@ViewDescriptor("user-detail-view.xml")
@EditedEntityContainer("userDc")
open class UserDetailView : StandardDetailView<User>() {

    @ViewComponent
    private lateinit var usernameField: TypedTextField<String>

    @ViewComponent
    private lateinit var passwordField: PasswordField

    @ViewComponent
    private lateinit var confirmPasswordField: PasswordField

    @ViewComponent
    private lateinit var timeZoneField: ComboBox<String>

    @Autowired
    private lateinit var entityStates: EntityStates

    @Autowired
    private lateinit var messageBundle: MessageBundle

    @Autowired
    private lateinit var passwordEncoder: PasswordEncoder

    @Subscribe
    fun onInit(event: InitEvent) {
        timeZoneField.setItems(listOf(*TimeZone.getAvailableIDs()))
    }

    @Subscribe
    fun onInitEntity(event: InitEntityEvent<User>) {
        usernameField.isReadOnly = false
        passwordField.isVisible = true
        confirmPasswordField.isVisible = true
    }

    @Subscribe
    fun onReady(event: ReadyEvent) {
        if (entityStates.isNew(editedEntity)) {
            usernameField.focus()
        }
    }

    @Subscribe
    fun onValidation(event: ValidationEvent) {
        if (entityStates.isNew(editedEntity)
                && !Objects.equals(passwordField.value, confirmPasswordField.value)) {
            event.errors.add(messageBundle.getMessage("passwordsDoNotMatch"))
        }
    }

    @Subscribe
    protected fun onBeforeSave(event: BeforeSaveEvent) {
        if (entityStates.isNew(editedEntity)) {
            editedEntity.password = passwordEncoder.encode(passwordField.value)
        }
    }
}